### PR TITLE
Communication tools: SPI, I2C, UART.

### PR DESCRIPTION
**Pull request by teoperisanu**
Add SPI, I2C, UART bitbanging and CLI tools.

### CODE REVIEW

File: tools/communication/src/spi.cpp [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+		trigger->setDigitalCondition(m2KSpiDesc->clock, libm2k::analog::NO_TRIGGER);
+		trigger->setDigitalCondition(m2KSpiDesc->mosi, libm2k::analog::NO_TRIGGER);
+		trigger->setDigitalCondition(m2KSpiDesc->miso, libm2k::analog::NO_TRIGGER);
+		trigger->setDigitalCondition(spiDesc->chip_select, libm2k::analog::NO_TRIGGER);
+		trigger->setDigitalMode(libm2k::digital::DIO_OR);
+		trigger->setDigitalDelay(0);
+
+		//set the state of CS and CLK to idle
+		m2KSpiDesc->digital->setValueRaw(spiDesc->chip_select, libm2k::digital::HIGH);
+		if (spiDesc->mode >> 1u) {
```
Comment (adisuciu) [inline tools/communication/src/spi.cpp]: spiDesc->mode & SPI_CPOL would be more descriptive

File: tools/communication/src/spi.cpp [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+			   uint8_t *data,
+			   uint8_t bytes_number)
+{
+	try {
+		auto *m2KSpiDesc = (m2k_spi_desc *) desc->extra;
+		//start reading - wait until buffer is pushed
+		std::thread thread_read(read, desc, data, bytes_number);
+		std::vector<unsigned short> buffer = createBuffer(desc, data, bytes_number);
+		//make sure the reading thread is waiting
+		std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
Comment (adisuciu) [inline tools/communication/src/spi.cpp]: 100ms seems like a long delay to arm the trigger & prepare the buffer.
@AlexandraTrifan mentioned errno from libiio could be leveraged here.

File: tools/communication/src/i2c.cpp [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+		if (operation) {
+			//read operation
+			setBit(address, 0);
+		}
+		//write address
+		writeByte(desc, buffer, address);
+		//write acknowledge
+		writeBit(desc, buffer, true);
+	} else {
+		address = 0x1E << 3u;
```
Comment (adisuciu) [inline tools/communication/src/i2c.cpp]: Provide #define for 0x1E

File: tools/m2kcli/commands/communication/uart.cpp [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+	std::string format;
+	if (arguments.count("format")) {
+		Validator::validate(arguments["format"], "format", format);
+	}
+
+	uartDesc->device_id = device;
+	std::vector<uint8_t> data(bytesNumber);
+	uart_read(uartDesc, data.data(), bytesNumber);
+
+	if (format.empty() || format == "text") {
```
Comment (adisuciu) [inline tools/m2kcli/commands/communication/uart.cpp]: Is it possible to create an option/format that gets data from stdin ?

File: tools/communication/src/spi.cpp (hunk 1 above)
Comment (teoperisanu) [reply tools/communication/src/spi.cpp]: Modified in commit cd7956b.

File: tools/communication/src/spi.cpp (hunk 2 above)
Comment (teoperisanu) [reply tools/communication/src/spi.cpp]: The sleeping time was reduced to 1 ms.

File: tools/communication/src/i2c.cpp (hunk 3 above)
Comment (teoperisanu) [reply tools/communication/src/i2c.cpp]: Modified in commit 59fafa3.

File: tools/m2kcli/commands/communication/uart.cpp (hunk 4 above)
Comment (teoperisanu) [reply tools/m2kcli/commands/communication/uart.cpp]: If the data content is empty (e.g. "--write date=") the data will be read from stdin.

**Comment by commodo** [comment]
This looks like too much code for one Pull-Request.
Maybe split it in 4-5 PRs?

Thing is:
* you won't expect/get a proper review on such a big changeset
* some changes that may be good, would block other changes that are open for debathe
* multiple PRs means that discussions can go in parallel on multiple discussion-threads [ or just threads ]

I typically recommend to people to try to do code-reviews in order to understand how to submit code for reviews.

**Pull request closed without merging**